tasks/main.yml

```diff
@@ -1,39 +1,138 @@
 ---
-# Put the tasks for your role here.
-
-# An example of how to set distribution and version specific internal variables
-# (defined in vars/ directory)
-# see https://github.com/oasis-roles/meta_standards#platform-specific-variables
 - name: Set platform/version specific variables
-  include_vars: "{{ __template_vars_file }}"
+  include_vars: "{{ __vpn_vars_file }}"
   loop:
     - "{{ ansible_facts['os_family'] }}.yml"
     - "{{ ansible_facts['distribution'] }}.yml"
-    - "{{ ansible_facts['distribution'] ~ '_' ~
-          ansible_facts['distribution_major_version'] }}.yml"
-    - "{{ ansible_facts['distribution'] ~ '_' ~
-          ansible_facts['distribution_version'] }}.yml"
+    - >-
+      {{ ansible_facts['distribution'] ~ '_' ~
+      ansible_facts['distribution_major_version'] }}.yml
+    - >-
+      {{ ansible_facts['distribution'] ~ '_' ~
+      ansible_facts['distribution_version'] }}.yml
   vars:
-    __template_vars_file: "{{ role_path }}/vars/{{ item }}"
-  when: __template_vars_file is file
+    __vpn_vars_file: "{{ role_path }}/vars/{{ item }}"
+  when: __vpn_vars_file is file
 
-# Examples of some tasks:
 - name: Ensure required packages are installed
   package:
-    name: "{{ __template_packages }}"
+    name: "{{ __vpn_packages }}"
     state: present
 
-- name: Ensure required services are enabled and started
-  service:
-    name: "{{ item }}"
-    state: started
-    enabled: yes
-  loop: "{{ __template_services }}"
-
-- name: Generate /etc/{{ __template_foo_config }}
-  template:
-    src: "{{ __template_foo_config }}.j2"
-    dest: /etc/{{ __template_foo_config }}
-    backup: yes
-  notify: template handler to restart services
+- name: Controller tasks
+  block:
+    - name: Enforce default auth method as needed
+      set_fact:
+        vpn_connections: |
+          {% for tunnel in vpn_connections %}
+          {%   set _ = tunnel.__setitem__(
+            "auth_method", tunnel.auth_method | d(vpn_auth_method)
+            ) %}
+          {% endfor %}
+          {{ vpn_connections }}
+
+    - name: Make sure that the hosts list is not empty
+      vars:
+        failure: >-
+          {%- for tunnel in vpn_connections %}
+          {%   if not tunnel.hosts -%}True{%- endif %}
+          {% endfor %}
+      fail:
+        msg: list of hosts is empty for one or more tunnels
+      when: '"True" in failure'
+
+    - name: Ensure cert_names are populated when auth_method is cert
+      vars:
+        failure: >-
+          {%- for tunnel in vpn_connections %}
+          {%   if tunnel.auth_method == 'cert' %}
+          {%     if tunnel.hosts is mapping %}
+          {%       for key, value in tunnel.hosts.items() %}
+          {% if value is none or 'cert_name' not in value -%}True{%-endif %}
+          {%       endfor %}
+          {%     else -%}True{%-endif %}
+          {%   endif %}
+          {% endfor %}
+      fail:
+        msg: cert_name is empty for one or more hosts in a tunnel
```

Comment:
It is indeed trivial, but better to be `cert_name is missing or empty for one or more hosts in a tunnel`?

BTW, I'm curious this check really captures an "empty" case? `cert_name` is in the `value`...
```
      - hosts:
          host1.example.com:
            cert_name: ""
```

Reply:
Correct, it wasn't capturing all the possible failure cases. Just fixed this in the [latest commit](https://github.com/linux-system-roles/vpn/pull/3/commits/aafdccc0a5187617674882ffb40eff78ac77e2ef). Thanks.